### PR TITLE
Patch for SAMOA-58 (Samoa AvroFileStream from HDFSFileStreamSource stops at end of first file)

### PR DESCRIPTION
**Pull request by edi-bice**
FileStreamSource seemed to support multiple files but during my testing it turned out otherwise - Samoa AvroFileStream from HDFSFileStreamSource stops at end of first file. I had to change AvroFileStream, ArffFileStream and their parent FileStream in order to make this work.

See following JIRA for additional detail:

https://issues.apache.org/jira/browse/SAMOA-58

Additionally, I modified bin/samoa, pom.xml, SystemUtils (as well as added a resource) to fix reading from HDFS on my cluster.

A seemingly unrelated change is the explicit test for supported Avro types so as to filter out any fields that are not supported instead of assuming all non-nominal (non-enum) fields are numeric and failing during reading.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -136,7 +136,7 @@
         <metrics-core.version>2.2.0</metrics-core.version>
         <miniball.version>1.0.3</miniball.version>
         <s4.version>0.6.0-incubating</s4.version>
-        <samza.version>0.7.0</samza.version>
+        <samza.version>0.10.1-MINE</samza.version>
```
Comment (gdfm) [inline pom.xml]: Probably should be 0.10.0

File: pom.xml [hunk 2]
```diff
@@ -167,10 +167,10 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
-                <version>3.1</version>
+                <version>3.5</version>
                 <configuration>
-                    <source>1.7</source>
-                    <target>1.7</target>
+                    <source>1.8</source>
```
Comment (gdfm) [inline pom.xml]: I don't think we want to move the source version to 1.8 unless we move the whole infrastructure (e.g., JDK) to Java 8

File: samoa-api/src/main/java/org/apache/samoa/evaluation/F1ClassificationPerformanceEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+import org.apache.samoa.moa.core.Measurement;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Vector;
+
+/**
+ * Created by Edi Bice (edi.bice gmail com) on 2/22/2016.
+ */
+public class F1ClassificationPerformanceEvaluator extends AbstractMOAObject implements
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/evaluation/F1ClassificationPerformanceEvaluator.java]: What does this class do that org.apache.samoa.evaluation.measures.F1 does not?

File: samoa-api/src/main/java/org/apache/samoa/moa/core/Measurement.java [hunk 4]
```diff
@@ -36,14 +36,19 @@
   private static final long serialVersionUID = 1L;
 
   protected String name;
-
   protected double value;
+  protected int fractionDigits;
 
   public Measurement(String name, double value) {
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/moa/core/Measurement.java]: This constructor should call the more general one with a default value (3 in this case)

File: samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java [hunk 5]
```diff
@@ -46,8 +46,8 @@
   private static final Logger logger = LoggerFactory.getLogger(AvroFileStream.class);
 
   public FileOption avroFileOption = new FileOption("avroFile", 'f', "Avro File(s) to load.", null, null, false);
-  public IntOption classIndexOption = new IntOption("classIndex", 'c',
-      "Class index of data. 0 for none or -1 for last attribute in file.", -1, -1, Integer.MAX_VALUE);
+  /*public IntOption classIndexOption = new IntOption("classIndex", 'c',
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java]: Don't we need the class index in the Avro file?

File: samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java [hunk 6]
```diff
@@ -140,7 +140,8 @@ protected boolean readNextInstanceFromFile() {
   public void prepareForUseImpl(TaskMonitor monitor, ObjectRepository repository) {
     super.prepareForUseImpl(monitor, repository);
     String filePath = this.avroFileOption.getFile().getAbsolutePath();
-    this.fileSource.init(filePath, AvroFileStream.AVRO_FILE_EXTENSION);
+    //this.fileSource.init(filePath, AvroFileStream.AVRO_FILE_EXTENSION);
+    this.fileSource.init(filePath, null);
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java]: Could you explain this change?

File: samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java [hunk 7]
```diff
@@ -40,11 +42,12 @@
   public FileOption arffFileOption = new FileOption("arffFile", 'f',
       "ARFF File(s) to load.", null, null, false);
 
-  public IntOption classIndexOption = new IntOption("classIndex", 'c',
+  /*public IntOption classIndexOption = new IntOption("classIndex", 'c',
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java]: Same as above, don't we need the class index anymore?

File: samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java [hunk 8]
```diff
@@ -57,32 +60,39 @@ public void prepareForUseImpl(TaskMonitor monitor, ObjectRepository repository)
   @Override
   protected void reset() {
     try {
-      if (this.fileReader != null)
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java]: Shouldn't we try to close open file readers?

File: samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java [hunk 9]
```diff
@@ -57,32 +60,39 @@ public void prepareForUseImpl(TaskMonitor monitor, ObjectRepository repository)
...
   @Override
-  protected boolean getNextFileReader() {
-    boolean ret = super.getNextFileReader();
-    if (ret) {
-      this.instances = new Instances(this.fileReader, 1, -1);
-      if (this.classIndexOption.getValue() < 0) {
-        this.instances.setClassIndex(this.instances.numAttributes() - 1);
-      } else if (this.classIndexOption.getValue() > 0) {
-        this.instances.setClassIndex(this.classIndexOption.getValue() - 1);
+  protected boolean getNextFileStream() {
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java]: I guess you try to close each file in the collection separately?
Does this work for a single file?

File: samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java [hunk 10]
```diff
@@ -20,12 +20,7 @@
  * #L%
  */
 
-import java.io.BufferedReader;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
-import java.io.Reader;
-
+import com.github.javacliparser.*;
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: We tend to prefer qualified imports to wildcard imports

File: samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java [hunk 11]
```diff
@@ -52,9 +49,18 @@
       's', "Source Type (HDFS, local FS)", FileStreamSource.class,
       "LocalFileStreamSource");
 
+  public IntOption classIndexOption = new IntOption("classIndex", 'c',
+          "Class index of data. 0 for none or -1 for last attribute in file.", -1, -1, Integer.MAX_VALUE);
+
+  private FloatOption floatOption = new FloatOption("classWeight", 'w', "", 1.0);
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: What is the idea behind class weights? Is it for multi class classification?

File: samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java [hunk 12]
```diff
@@ -90,6 +105,13 @@ public InstanceExample nextInstance() {
       readNextInstanceFromStream();
     }
     InstanceExample prevInstance = this.getLastInstanceRead();
+    if (classWeights != null && classWeights.length > 0) {
+      int i = (int) prevInstance.instance.classValue();
+      double w = 1.0;
+      if (i>=0 && i<classWeights.length)
```
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: Please put spaces around operators as per coding conventions

File: samoa-instances/src/main/java/org/apache/samoa/instances/AvroLoader.java [hunk 13]
```diff
@@ -254,12 +257,30 @@ protected InstanceInformation getHeader() {
         List<String> attributeLabels = attributeSchema.getEnumSymbols();
         attributes.add(new Attribute(field.name(), attributeLabels));
       }
-      else
+      else if (isNumeric(field))
```
Comment (gdfm) [inline samoa-instances/src/main/java/org/apache/samoa/instances/AvroLoader.java]: I think the idea here is that if we don't know better, we treat the attribute as numeric.
With this change, we ignore attributes we don't know how to treat.
Is my understanding correct?

**Comment by gdfm** [comment]
Thanks for the patch @edi-bice !
There are a few changes that go beyond the scope of the patch, and their goal is not entirely clear to me.
For the rest, apart from some small doubts, the patch looks good.
Please have a look at my comments, and see if they make sense.
I'd like to get the patch in soon.

File: pom.xml (hunk 1 above)
Comment (edi-bice) [inline pom.xml]: Yes, you're correct - my work environment details crept in as I'm getting used to the whole process.

File: pom.xml (hunk 2 above)
Comment (edi-bice) [inline pom.xml]: Agree.

File: samoa-api/src/main/java/org/apache/samoa/evaluation/F1ClassificationPerformanceEvaluator.java (hunk 3 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/evaluation/F1ClassificationPerformanceEvaluator.java]: I'm using Samoa for extremely unbalanced classification and needed precision, recall, F-measure along with supports. measures.F1 seemed to apply to clustering only (and I didn't know how to use in my scenario) though I can't see why we couldn't in theory use one F1 measure for both multiclass classification and clustering.

File: samoa-api/src/main/java/org/apache/samoa/moa/core/Measurement.java (hunk 4 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/moa/core/Measurement.java]: changed to have the two parameter constructor above call the three parameter one defaulting third parameter fractionDigits to 3

File: samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java (hunk 5 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java]: Yes we do but since we also need in ArffFileStream, I moved up to the common FileStream parent.

File: samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java (hunk 6 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/moa/streams/AvroFileStream.java]: My Avro files do not have extensions. I realize this is not my personal software but it did happen to me and it can happen to others. Why require an extension at all?

File: samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java (hunk 7 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java]: Same as above, it's being inherited from parent.

File: samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java (hunk 9 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java]: Good point. I'm not sure if there are tests that cover both scenarios. But at least it seems from code that the file reader/stream is closed (and set to null) upon failure to read an instance (end of file) and is also checked and closed upon call to getNextFileStream.

File: samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java (hunk 10 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: You got it.

File: samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java (hunk 11 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: Imbalanced classification (for example in payments fraud prevention 1 fraud to hundreds/thousands good transactions) requires class weights to force classifier.

File: samoa-instances/src/main/java/org/apache/samoa/instances/AvroLoader.java (hunk 13 above)
Comment (edi-bice) [inline samoa-instances/src/main/java/org/apache/samoa/instances/AvroLoader.java]: You are correct and hopefully agree that it is an improvement as it is now accepting of more data sets (including mine with has way too many string attributes - could add a pre-processing step/job before Samoa but thought this nice to have regardless).

**Comment by edi-bice** [comment]
@gdfm apologies for the delay - was away at a conference for a week.

Thanks for reviewing my patch and being receptive to the changes. See my comments for answers to your questions. Will push the changes you've asked for hopefully in a couple of hours.

File: samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java (hunk 8 above)
Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/streams/ArffFileStream.java]: I now call filesource.reset() which closes the filestream.

**Comment by edi-bice** [comment]
Pushed the changes you requested @gdfm

File: samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java (hunk 11 above)
Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: Would instance weights work in this case?
Could you summarize the difference between class weights and instance weights?

Comment (edi-bice) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: Yes, it would. The class weight option is indeed implemented via instance weights. 

Most machine learning algorithms focus on total error and in extremely imbalanced scenarios (fraud, terrorism, disease) would fail to detect the sparse class which is really what we're after. Class weighting allows one to incorporate apriori knowledge of the imbalance. For example sklearn, R e1071 SVM packages have class weights options.

Comment (gdfm) [inline samoa-api/src/main/java/org/apache/samoa/streams/FileStream.java]: I see your point. So class weight is an easy way to set instance weight for all the instances that share the same class.
Looks good to me!

**Comment by gdfm** [comment]
+1 and thanks for the contribution!

**Comment by gdfm** [comment]
@abifet @nicolas-kourtellis could you also review the patch?

**Comment by nicolas-kourtellis** [comment]
+1
I didn't run the Avro fix, but the rest of the additions look good. Thanks edi-bice!

Note: What I am not sure is if all these additions should have been committed as a single patch, or independent ones. The patch covers several different things and if someone wants to roll out one of them, it may not be easy..

**Comment by abifet** [comment]
+1 I like the idea of having different weights for different classes. The F1ClassificationPerformanceEvaluator is giving also recall and precision, that's a nice contribution. We may think in the future to add a windowing version of it. Thanks @edi-bice!

**Comment by edi-bice** [comment]
I agree with you @nicolas-kourtellis about the patch being multiple seemingly unrelated things. Typical rookie mistake - took off on my fork making all the changes I needed to get things working for me. Didn't squash either. Had it not been so long since I last worked on this, I'd offer to start splitting things. 

@abifet let me know what the windowing version should look like and I can work on it.

**Comment by abifet** [comment]
@edi-bice  It's like WindowClassificationPerformanceEvaluator. This is a windowing version of BasicClassificationPerformanceEvaluator that takes into account the performance on the last instances, instead of all the instances from the beginning of the stream.
